runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java

```diff
@@ -460,9 +460,15 @@ public void stop(boolean shutdown) {
         final CompletableFuture<T> cfTimeout =
             CFUtils.within(cfBenchmarked, Duration.ofMillis(timeoutResponse));
         cfTimeout.exceptionally(e -> {
-            outstandingRequests.remove(thisRequest);
-            log.debug("Remove request {} to {} due to timeout! Message:{}",
-                    thisRequest, node, message);
+            // CFUtils.within() can wrap different kinds of exceptions in
+            // CompletionException, just dealing with TimeoutException here since
+            // the router is not aware of it and this::completeExceptionally()
+            // takes care of others. This avoids handling same exception twice.
+            if (e.getCause() instanceof TimeoutException) {
+                outstandingRequests.remove(thisRequest);
+                log.debug("sendMessageAndGetCompletable: Remove request {} to {} due to timeout! Message:{}",
```

Comment:
Is it possible to have an exception that prevents the TimeoutException to happen, which results in leaking requests in outstandingRequests?

Reply:
In case of exceptions other than TimeoutException, `completeExceptionally()` will handle it and remove it from outstandingRequests. https://github.com/CorfuDB/CorfuDB/blob/1d6c4a1a70f91b5315fa6162a7cdce558d4bb339/runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java#L530